ckanext/search/filters.py

```diff
@@ -55,7 +62,56 @@ def _check_filter_operator(key: str, value: Any) -> None:
         raise Invalid("Filter operations must be defined as a list of dicts")
 
 
-def _process_filter_operator_members(value: Any) -> List[FilterOp]:
+def _combine_filter_operator_members(
+    value: Any, parent_operator: str
+) -> List[FilterOp]:
+    """
+    Combine child operator members of the same type, e.g.:
+
+    {
+        "$and": [
+            {"field1": "value1"},
+            {
+                "$and": [
+                    {"field2": "value2"},
+                    {"$and": [
+                        {"field3": "value3"},
+                        {"field4": "value4"}
+                        ]
+                    },
+                ]
+            },
+        ]
+    }
+
+    will generate a single $and filter:
+
+        FilterOp(
+            field=None,
+            op="$and",
+            value=[
+                FilterOp(field="field1", op="eq", value="value1"),
+                FilterOp(field="field2", op="eq", value="value2"),
+                FilterOp(field="field3", op="eq", value="value3"),
+                FilterOp(field="field4", op="eq", value="value4"),
+            ],
+        )
+    """
+    child_filters = _process_filter_operator_members(value, parent_operator)
+
+    combined_filters = []
+    for filter_ in child_filters:
+        if filter_.op == parent_operator and parent_operator in COMBINE_OPERATORS:
```

Comment:
Is "commutative" the right term for this kind of operation? When plugins register operators we can let them declare if these new operators can be combined/flattened in the same way too.

Reply:
If I remember correctly from my school days commutative relates to the order of operands (e.g. A + B = B + A)